File: lib/fluxsource/a2rfluxsource.cc
#include "lib/core/globals.h"
#include "lib/data/fluxmap.h"
#include "lib/fluxsource/fluxsource.pb.h"
#include "lib/fluxsource/fluxsource.h"
#include "lib/config/proto.h"
#include <fstream>

struct A2Rv2Flux
{
    std::vector<Bytes> flux;
    nanoseconds_t index;
};

class A2rv2FluxSourceIterator : public FluxSourceIterator
{
public:
    A2rv2FluxSourceIterator(A2Rv2Flux& flux): _flux(flux) {}

    bool hasNext() const override
    {
        return _count != _flux.flux.size();
    }

    std::unique_ptr<const Fluxmap> next() override
    {
        nanoseconds_t index = _flux.index;
        Bytes& asbytes = _flux.flux[_count++];
        ByteReader br(asbytes);

        auto fluxmap = std::make_unique<Fluxmap>();
        while (!br.eof())
        {
            unsigned aticks = 0;
            for (;;)
            {
                unsigned i = br.read_8();
                aticks += i;
                if (i != 0xff)
                    break;
            }

            nanoseconds_t interval = aticks * 125;
            if ((index >= 0) && (index < interval))
            {
                fluxmap->appendInterval(index);
                fluxmap->appendIndex();
                interval -= index;
            }
            index -= interval;

            fluxmap->appendInterval(interval / NS_PER_TICK);
            fluxmap->appendPulse();
        }

        return fluxmap;
    }

private:
    A2Rv2Flux& _flux;
    int _count = 0;
};

class A2rFluxSource : public FluxSource
{
public:
    A2rFluxSource(const A2rFluxSourceProto& config): _config(config)
    {
        _data = Bytes::readFromFile(_config.filename());
        ByteReader br(_data);

        switch (br.read_be32())
        {
            case 0x41325232:
            {
                _version = 2;
                Bytes info = findChunk("INFO");
                int disktype = info[33];
                if (disktype == 1)
                {
                    /* 5.25" with quarter stepping. */
                    _extraConfig.mutable_drive()->set_tracks(160);
                    _extraConfig.mutable_drive()->set_heads(1);
                    _extraConfig.mutable_drive()->set_drive_type(
                        DRIVETYPE_APPLE2);
                }
                else
                {
                    /* 3.5". */
                    _extraConfig.mutable_drive()->set_drive_type(
                        DRIVETYPE_80TRACK);
                }

                Bytes stream = findChunk("STRM");
                ByteReader bsr(stream);
                for (;;)
                {
                    int location = bsr.read_8();
                    if (location == 0xff)
                        break;
                    auto key = (disktype == 1) ? std::make_pair(location, 0)
                                               : std::make_pair(location >> 1,
                                                     location & 1);

                    bsr.skip(1);
                    uint32_t len = bsr.read_le32();
                    nanoseconds_t index = (nanoseconds_t)bsr.read_le32() * 125;
                    auto it = _v2data.find(key);
                    if (it == _v2data.end())
                    {
                        _v2data[key] = std::make_unique<A2Rv2Flux>();
                        it = _v2data.find(key);
                        it->second->index = index;
                    }

                    it->second->flux.push_back(bsr.read(len));
                }

                break;
            }

            default:
                error("unsupported A2R version");
        }
    }

public:
    std::unique_ptr<FluxSourceIterator> readFlux(int track, int head) override
    {
        switch (_version)
        {
            case 2:
            {
                auto i = _v2data.find(std::make_pair(track, head));
                if (i != _v2data.end())
                    return std::make_unique<A2rv2FluxSourceIterator>(
                        *i->second);
                else
                    return std::make_unique<EmptyFluxSourceIterator>();
            }

            default:
                error("unsupported A2R version");
        }
    }

    void recalibrate() override {}

private:
    Bytes findChunk(Bytes id)
    {
        uint32_t offset = 8;
        while (offset < _data.size())
        {
            ByteReader br(_data);
            br.seek(offset);
            if (br.read(4) == id)
            {
                uint32_t size = br.read_le32();
                return br.read(size);
            }

            offset += br.read_le32() + 8;
        }

        error("A2R file missing chunk");
    }

private:
    const A2rFluxSourceProto& _config;
    Bytes _data;
    std::ifstream _if;
    int _version;
    std::map<std::pair<int, int>, std::unique_ptr<A2Rv2Flux>> _v2data;
};

std::unique_ptr<FluxSource> FluxSource::createA2rFluxSource(
    const A2rFluxSourceProto& config)
{
    return std::unique_ptr<FluxSource>(new A2rFluxSource(config));
}

File: java/com/cowlark/fluxengine/fluxsource/A2RFluxSource.java
package com.cowlark.fluxengine.fluxsource;

import com.cowlark.fluxengine.config.ConfigBuilder;
import com.cowlark.fluxengine.config.ConfigProto;
import com.cowlark.fluxengine.core.ByteReader;
import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.core.FluxEngineException;
import com.cowlark.fluxengine.data.CylinderHead;
import com.cowlark.fluxengine.data.Locations;
import com.cowlark.fluxengine.external.DriveType;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.TreeMap;

/**
 * A flux source which reads an A2R flux file, ported from
 * lib/fluxsource/a2rfluxsource.cc.
 */
public class A2RFluxSource extends FluxSource
{
    private final TreeMap<CylinderHead, A2Rv2Flux> v2data = new TreeMap<>();
    private final A2rFluxSourceProto config;
    private final Bytes data;
    protected ConfigProto extraConfig;
    private int version;

    public A2RFluxSource(A2rFluxSourceProto config)
    {
        this.config = config;
        data = readFile(config.getFilename());
        ByteReader br = new ByteReader(data);

        switch (br.readBe32())
        {
            case 0x41325232:
            {
                version = 2;
                Bytes info = findChunk(new Bytes("INFO"));
                int disktype = info.getByte(33) & 0xff;
                DriveType driveType;
                if (disktype == 1)
                {
                    /* 5.25" with quarter stepping. */
                    driveType = DriveType.DRIVETYPE_APPLE2;
                } else
                {
                    /* 3.5". */
                    driveType = DriveType.DRIVETYPE_80TRACK;
                }

                Bytes stream = findChunk(new Bytes("STRM"));
                ByteReader bsr = new ByteReader(stream);
                for (; ; )
                {
                    int location = bsr.read8();
                    if (location == 0xff)
                        break;
                    CylinderHead key = (disktype == 1) ?
                            new CylinderHead(location, 0) :
                            new CylinderHead(location >> 1, location & 1);

                    bsr.skip(1);
                    int len = bsr.readLe32();
                    double index = (double) bsr.readLe32() * 125;
                    A2Rv2Flux entry = v2data.get(key);
                    if (entry == null)
                    {
                        entry = new A2Rv2Flux();
                        entry.index = index;
                        v2data.put(key, entry);
                    }

                    entry.flux.add(bsr.read(len));
                }

                List<CylinderHead> chs = new ArrayList<>(v2data.keySet());

                ConfigProto.Builder builder = ConfigProto.newBuilder();
                builder.getDriveBuilder().setDriveType(driveType);
                builder.getDriveBuilder().setTracks(Locations.convertCylinderHeadsToString(chs));
                extraConfig = builder.build();
                break;
            }

            default:
                error("unsupported A2R version");
        }
    }

    private static Bytes readFile(String filename)
    {
        try
        {
            return new Bytes(Files.readAllBytes(Path.of(filename)));
        } catch (IOException e)
        {
            throw new FluxEngineException(
                    "cannot open input file '" + filename + "': " + e.getMessage());
        }
    }

    private static void error(String message)
    {
        throw new FluxEngineException(message);
    }

    @Override
    public void adjustConfig(ConfigBuilder configBuilder)
    {
        configBuilder.mergeConfig(extraConfig);
    }

    @Override
    public FluxSourceIterator readFlux(FluxReadParameters parameters)
    {
        switch (version)
        {
            case 2:
            {
                A2Rv2Flux entry =
                        v2data.get(new CylinderHead(parameters.cylinder(), parameters.head()));
                if (entry != null)
                    return new A2RFluxSourceIterator(entry.flux, entry.index);
                else
                    return new EmptyFluxSourceIterator();
            }

            default:
                error("unsupported A2R version");
                return null;
        }
    }

    @Override
    public void recalibrate()
    {
    }

    private Bytes findChunk(Bytes id)
    {
        long offset = 8;
        while (offset < data.size())
        {
            ByteReader br = new ByteReader(data);
            br.seek((int) offset);
            if (br.read(4).equals(id))
            {
                int size = br.readLe32();
                return br.read(size);
            }

            offset += (long) br.readLe32() + 8;
        }

        error("A2R file missing chunk");
        return null;
    }

    static class A2Rv2Flux
    {
        List<Bytes> flux = new ArrayList<>();
        double index;
    }
}